### PR TITLE
Dashboard QP endpoints should ignore :default values and treat them as no value and compile Field filters to `1 = 1`

### LINKED ISSUE #20493: Dashboard filter with defaults, doesn't work if values are removed, query fails
**Describe the bug**
Dashboard filter with defaults, doesn't work if values are removed, query fails.
Feels like #20438 is related, but very different error.

**To Reproduce**
1. Native > Sample Dataset > set variable to Field Filter of Products.Category and save as "Q1"
`SELECT COUNT(*) FROM "PRODUCTS" WHERE {{cat}}`
2. Add "Q1" to dashboard, and add filter and set a default value - save dashboard
image
3. Remove the widget value - query will fail
image

<details><summary>Full stacktrace</summary>

```
2022-02-14 16:44:36,487 ERROR middleware.catch-exceptions :: Error processing query: null
{:database_id 4,
 :started_at #t "2022-02-14T16:44:36.232219+01:00[Europe/Copenhagen]",
 :via
 [{:status :failed,
   :class clojure.lang.ExceptionInfo,
   :error "Unable to substitute parameters: No values provided for operator: :string/=",
   :stacktrace
   ["--> driver.sql.parameters.substitute$substitute$fn__78003.invoke(substitute.clj:83)"
    "driver.sql.parameters.substitute$substitute.invokeStatic(substitute.clj:80)"
    "driver.sql.parameters.substitute$substitute.invoke(substitute.clj:71)"
    "driver.sql$fn__78521$substitute_native_parameters__78514__78526$fn__78530.invoke(sql.clj:46)"
    "driver.sql$fn__78521$substitute_native_parameters__78514__78526.invoke(sql.clj:42)"
    "query_processor.middleware.parameters.native$expand_inner.invokeStatic(native.clj:39)"
    "query_processor.middleware.parameters.native$expand_inner.invoke(native.clj:30)"
    "query_processor.middleware.parameters$expand_one.invokeStatic(parameters.clj:50)"
    "query_processor.middleware.parameters$expand_one.invoke(parameters.clj:41)"
    "query_processor.middleware.parameters$expand_all$replace_48586__48587.invoke(parameters.clj:59)"
    "mbql.util.match.impl$replace_in_collection$iter__27401__27405$fn__27406.invoke(impl.cljc:44)"
    "mbql.util.match.impl$replace_in_collection.invokeStatic(impl.cljc:43)"
    "mbql.util.match.impl$replace_in_collection.invoke(impl.cljc:38)"
    "query_processor.middleware.parameters$expand_all$replace_48586__48587.invoke(parameters.clj:59)"
    "query_processor.middleware.parameters$expand_all.invokeStatic(parameters.clj:59)"
    "query_processor.middleware.parameters$expand_all.invoke(parameters.clj:53)"
    "query_processor.middleware.parameters$expand_all.invokeStatic(parameters.clj:56)"
    "query_processor.middleware.parameters$expand_all.invoke(parameters.clj:53)"
    "query_processor.middleware.parameters$expand_parameters.invokeStatic(parameters.clj:77)"
    "query_processor.middleware.parameters$expand_parameters.invoke(parameters.clj:73)"
    "query_processor.middleware.parameters$fn__48602$substitute_parameters_STAR___48607$fn__48608.invoke(parameters.clj:82)"
    "query_processor.middleware.parameters$fn__48602$substitute_parameters_STAR___48607.invoke(parameters.clj:79)"
    "query_processor.middleware.parameters$substitute_parameters$fn__48628.invoke(parameters.clj:111)"
    "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__46025.invoke(resolve_referenced.clj:79)"
    "query_processor.middleware.expand_macros$expand_macros$fn__52249.invoke(expand_macros.clj:184)"
    "query_processor.middleware.add_timezone_info$add_timezone_info$fn__48407.invoke(add_timezone_info.clj:15)"
    "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__51200.invoke(splice_params_in_response.clj:32)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__50647$fn__50652.invoke(resolve_database_and_driver.clj:35)"
    "driver$do_with_driver.invokeStatic(driver.clj:60)"
    "driver$do_with_driver.invoke(driver.clj:56)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__50647.invoke(resolve_database_and_driver.clj:34)"
    "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__46571.invoke(fetch_source_query.clj:286)"
    "query_processor.middleware.store$initialize_store$fn__46762$fn__46763.invoke(store.clj:11)"
    "query_processor.store$do_with_store.invokeStatic(store.clj:44)"
    "query_processor.store$do_with_store.invoke(store.clj:38)"
    "query_processor.middleware.store$initialize_store$fn__46762.invoke(store.clj:10)"
    "query_processor.middleware.validate$validate_query$fn__50982.invoke(validate.clj:10)"
    "query_processor.middleware.normalize_query$normalize$fn__50989.invoke(normalize_query.clj:22)"
    "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__48353.invoke(add_rows_truncated.clj:35)"
    "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__49645.invoke(results_metadata.clj:82)"
    "query_processor.middleware.constraints$add_default_userland_constraints$fn__48371.invoke(constraints.clj:42)"
    "query_processor.middleware.process_userland_query$process_userland_query$fn__50923.invoke(process_userland_query.clj:146)"
    "query_processor.middleware.catch_exceptions$catch_exceptions$fn__51280.invoke(catch_exceptions.clj:169)"
    "query_processor.reducible$async_qp$qp_STAR___43323$thunk__43324.invoke(reducible.clj:103)"
    "query_processor.reducible$async_qp$qp_STAR___43323$fn__43326.invoke(reducible.clj:108)"],
   :error_type :invalid-parameter,
   :ex-data
   {:type :invalid-parameter,
    :params
    {"cat"
     (metabase.driver.common.parameters/map->FieldFilter
      {:field
       {:description "The type of product, valid values include: Doohicky, Gadget, Gizmo and Widget",
        :database_type "VARCHAR",
        :semantic_type :type/Category,
        :table_id 10,
        :coercion_strategy nil,
        :name "CATEGORY",
        :settings nil,
        :effective_type :type/Text,
        :parent_id nil,
        :id 107,
        :visibility_type :normal,
        :display_name "Category",
        :fingerprint
        {:global {:distinct-count 4, :nil% 0.0},
         :type
         {:type/Text
          {:percent-json 0.0, :percent-url 0.0, :percent-email 0.0, :percent-state 0.0, :average-length 6.375}}},
        :base_type :type/Text},
       :value {:type :string/=, :id "192414c3", :default ["Doohickey"]}})},
    :parsed-query ["SELECT COUNT(*) FROM \"PRODUCTS\" WHERE " (metabase.driver.common.parameters/->Param "cat")]}}],
 :error_type :invalid-parameter,
 :json_query
 {:constraints {:max-results 10000, :max-results-bare-rows 2000},
  :type :native,
  :middleware {:js-int-to-string? true, :ignore-cached-results? false},
  :native
  {:template-tags
   {"cat"
    {:id "6e5503c4-6177-1a30-92cf-4dedeb33ac10",
     :name "cat",
     :display-name "Cat",
     :type :dimension,
     :dimension [:field 107 nil],
     :widget-type :category,
     :default nil}},
   :query "SELECT COUNT(*) FROM \"PRODUCTS\" WHERE {{cat}}"},
  :database 4,
  :parameters [{:type :string/=, :id "192414c3", :default ["Doohickey"], :target [:dimension [:template-tag "cat"]]}],
  :async? true,
  :cache-ttl nil},
 :status :failed,
 :class clojure.lang.ExceptionInfo,
 :stacktrace
 ["--> driver.common.parameters.operators$fn__48269$verify_type_and_arity__48274$fn__48275.invoke(operators.clj:43)"
  "driver.common.parameters.operators$fn__48269$verify_type_and_arity__48274.invoke(operators.clj:24)"
  "driver.common.parameters.operators$fn__48296$to_clause__48301$fn__48317.invoke(operators.clj:59)"
  "driver.common.parameters.operators$fn__48296$to_clause__48301.invoke(operators.clj:55)"
  "driver.sql.parameters.substitution$fn__77735$field_filter__GT_replacement_snippet_info__77740$fn__77746.invoke(substitution.clj:267)"
  "driver.sql.parameters.substitution$fn__77735$field_filter__GT_replacement_snippet_info__77740.invoke(substitution.clj:256)"
  "driver.sql.parameters.substitution$fn__77766.invokeStatic(substitution.clj:301)"
  "driver.sql.parameters.substitution$fn__77766.invoke(substitution.clj:286)"
  "driver.sql.parameters.substitute$substitute_field_filter.invokeStatic(substitute.clj:14)"
  "driver.sql.parameters.substitute$substitute_field_filter.invoke(substitute.clj:9)"
  "driver.sql.parameters.substitute$substitute_param.invokeStatic(substitute.clj:31)"
  "driver.sql.parameters.substitute$substitute_param.invoke(substitute.clj:25)"
  "driver.sql.parameters.substitute$substitute_STAR_$fn__77994.invoke(substitute.clj:64)"
  "driver.sql.parameters.substitute$substitute_STAR_.invokeStatic(substitute.clj:57)"
  "driver.sql.parameters.substitute$substitute_STAR_.invoke(substitute.clj:54)"
  "driver.sql.parameters.substitute$substitute$fn__78003.invoke(substitute.clj:81)"
  "driver.sql.parameters.substitute$substitute.invokeStatic(substitute.clj:80)"
  "driver.sql.parameters.substitute$substitute.invoke(substitute.clj:71)"
  "driver.sql$fn__78521$substitute_native_parameters__78514__78526$fn__78530.invoke(sql.clj:46)"
  "driver.sql$fn__78521$substitute_native_parameters__78514__78526.invoke(sql.clj:42)"
  "query_processor.middleware.parameters.native$expand_inner.invokeStatic(native.clj:39)"
  "query_processor.middleware.parameters.native$expand_inner.invoke(native.clj:30)"
  "query_processor.middleware.parameters$expand_one.invokeStatic(parameters.clj:50)"
  "query_processor.middleware.parameters$expand_one.invoke(parameters.clj:41)"
  "query_processor.middleware.parameters$expand_all$replace_48586__48587.invoke(parameters.clj:59)"
  "mbql.util.match.impl$replace_in_collection$iter__27401__27405$fn__27406.invoke(impl.cljc:44)"
  "mbql.util.match.impl$replace_in_collection.invokeStatic(impl.cljc:43)"
  "mbql.util.match.impl$replace_in_collection.invoke(impl.cljc:38)"
  "query_processor.middleware.parameters$expand_all$replace_48586__48587.invoke(parameters.clj:59)"
  "query_processor.middleware.parameters$expand_all.invokeStatic(parameters.clj:59)"
  "query_processor.middleware.parameters$expand_all.invoke(parameters.clj:53)"
  "query_processor.middleware.parameters$expand_all.invokeStatic(parameters.clj:56)"
  "query_processor.middleware.parameters$expand_all.invoke(parameters.clj:53)"
  "query_processor.middleware.parameters$expand_parameters.invokeStatic(parameters.clj:77)"
  "query_processor.middleware.parameters$expand_parameters.invoke(parameters.clj:73)"
  "query_processor.middleware.parameters$fn__48602$substitute_parameters_STAR___48607$fn__48608.invoke(parameters.clj:82)"
  "query_processor.middleware.parameters$fn__48602$substitute_parameters_STAR___48607.invoke(parameters.clj:79)"
  "query_processor.middleware.parameters$substitute_parameters$fn__48628.invoke(parameters.clj:111)"
  "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__46025.invoke(resolve_referenced.clj:79)"
  "query_processor.middleware.expand_macros$expand_macros$fn__52249.invoke(expand_macros.clj:184)"
  "query_processor.middleware.add_timezone_info$add_timezone_info$fn__48407.invoke(add_timezone_info.clj:15)"
  "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__51200.invoke(splice_params_in_response.clj:32)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__50647$fn__50652.invoke(resolve_database_and_driver.clj:35)"
  "driver$do_with_driver.invokeStatic(driver.clj:60)"
  "driver$do_with_driver.invoke(driver.clj:56)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__50647.invoke(resolve_database_and_driver.clj:34)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__46571.invoke(fetch_source_query.clj:286)"
  "query_processor.middleware.store$initialize_store$fn__46762$fn__46763.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:44)"
  "query_processor.store$do_with_store.invoke(store.clj:38)"
  "query_processor.middleware.store$initialize_store$fn__46762.invoke(store.clj:10)"
  "query_processor.middleware.validate$validate_query$fn__50982.invoke(validate.clj:10)"
  "query_processor.middleware.normalize_query$normalize$fn__50989.invoke(normalize_query.clj:22)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__48353.invoke(add_rows_truncated.clj:35)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__49645.invoke(results_metadata.clj:82)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__48371.invoke(constraints.clj:42)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__50923.invoke(process_userland_query.clj:146)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__51280.invoke(catch_exceptions.clj:169)"
  "query_processor.reducible$async_qp$qp_STAR___43323$thunk__43324.invoke(reducible.clj:103)"
  "query_processor.reducible$async_qp$qp_STAR___43323$fn__43326.invoke(reducible.clj:108)"],
 :card_id 603,
 :context :dashboard,
 :error "No values provided for operator: :string/=",
 :row_count 0,
 :running_time 0,
 :ex-data {:param-type :string/=, :param-value nil, :field-id 107, :type :invalid-parameter},
 :data {:rows [], :cols []}}

2022-02-14 16:44:36,502 DEBUG middleware.log :: POST /api/dashboard/6/dashcard/1190/card/603/query 202 [ASYNC: completed] 272.2 ms (13 DB calls) App DB connections: 0/13 Jetty threads: 3/50 (3 idle, 0 queued) (106 total active threads) Queries in flight: 0 (0 queued)
```

</details>

**Information about your Metabase Installation:**
0.42.0

### PR DESCRIPTION
**Pull request by camsaul**
~It wasn't considering `:default` for the value and failing because the param had no value. Fix itself is only a few lines. The rest of this PR is additional logging and new tests~

Backend is not supposed to handle `:default` Dashboard parameter values. The FE client does this.

Fixes #20493

~This might also fixx #20438~

### CODE REVIEW

File: test/metabase/driver/sql/parameters/substitute_test.clj [hunk 1]
```diff
@@ -167,16 +167,15 @@
...
-
+                                                        :value value}})})))))))))
 
 ;;; -------------------------------------------- Referenced Card Queries ---------------------------------------------
 
 (deftest substitute-referenced-card-query-test
   (testing "Referenced card query substitution"
     (let [query ["SELECT * FROM " (param "#123")]]
       (is (= ["SELECT * FROM (SELECT 1 `x`)" []]
-             (substitute query {"#123" (i/map->ReferencedCardQuery {:card-id 123, :query "SELECT 1 `x`"})})))))))
```
Comment (camsaul) [inline test/metabase/driver/sql/parameters/substitute_test.clj:179]: Unrelated to this fix but this `deftest` was actually inside of the parent test form. Fixed the parens.

File: src/metabase/driver/common/parameters/values.clj [hunk 2]
```diff
@@ -81,8 +81,12 @@
...
   (let [matching-params  (params-with-target params [:dimension [:template-tag (:name tag)]])
         normalize-params (fn [params]
-                           ;; remove `:target` which is no longer needed after this point.
-                           (let [params (map #(dissoc % :target) params)]
+                           (let [params (for [param params]
+                                          ;; If param specifies a `:default` but no `:value`, use the default value
+                                          (merge (when-let [default (:default param)]
+                                                   {:value default})
+                                                 ;; remove `:target` which is no longer needed after this point.
+                                                 (dissoc param :target)))]
```
Comment (camsaul) [inline src/metabase/driver/common/parameters/values.clj:89]: This is the actual fix

**Review by dpsutton — APPROVED** [review]

**Review by flamber — CHANGES_REQUESTED** [review]
Tested `e72c729`. This changes the behavior and makes it impossible to remove dashboard defaults.

1. Setup a dashboard with default filter
2. Go to `/dashboard/:id` - it will automatically add the filter `/dashboard/:id?text=Doohickey`
3. Then when I clear the widget it changes to `/dashboard/:id?test=`, and it should now show all results, but with this PR it still applies the default value for the card
image

**Comment by camsaul** [comment]
@flamber are you **sure** that in older versions you could override the default value to somehow be nothing? If you look at the code I didn't change anything that would change that behavior and make it suddenly stop working where it had worked before.

**Comment by flamber** [comment]
@camsaul Try 0.41.6 - I would not block a PR if I didn't think it would cause more problems.

**Comment by nemanjaglumac** [comment]
@camsaul keep in mind that this issue was reproduced in https://github.com/metabase/metabase/pull/20495. I can update repro steps to also check for the thing Flamber found in his review https://github.com/metabase/metabase/pull/20503#pullrequestreview-882957652.

Let me know if you want me to do it in this branch (if you rebase it or include `master` in it), or do you want me to update it on the `master` first?

**Comment by camsaul** [comment]
Ok I'm a little confused on how this is actually supposed to work. I can tweak the code to ignore Dashboard `:default` parameter values (because the expectation seems to be that the FE client will explicitly pass those in as `:value` if it wants to use them) but then that breaks #19494.

In #19494 it seems that the clear expectation was that we **would** take default values into account on the backend

**Comment by nemanjaglumac** [comment]
@flamber can correct me because he has much better understanding, but dashboard would have the default parameter set on the visit. All good so far.

But you should also be able to remove that default filter value (set via filter widget) while exploring data and it should give you ALL results. The url changes to `/dashboard/:id?test=` as Flamber pointed out. Results on the page should reflect that.

On the next visit, the default filter would be set again.

**Comment by nemanjaglumac** [comment]
In other words, this:
https://www.loom.com/share/27047a48cba2435a9b32e817fa8b5dc6

**Comment by camsaul** [comment]
I think I've figured out how this is supposed to work. @flamber, can you try this out again and check #19494 as well to make sure that still works. @nemanjaglumac feel free to add more tests to this branch.

**Review by flamber — APPROVED** [review]
Tested `8c7a56b` and it works as expected (similar to 41.6) :tada: 
And I also re-tested #19494, which still works as expected.

EDIT: I've been testing it for the past 15 minutes, which is why the response is so fast

**Comment by nemanjaglumac** [comment]
>  @nemanjaglumac feel free to add more tests to this branch.

@camsaul The repro for this has already been merged to `master` previously. But since @flamber tested this manually and can confirm, please feel free to merge it and I'll enable the repro after you do that.

Wouldn't want to keep this PR any longer and to wait for one more CI cycle.

**Pull request merged**